db/schema_applier.cc

```diff
@@ -440,41 +442,43 @@ static aggregate_diff diff_aggregates_rows(const schema_result& aggr_before, con
     return {std::move(created), std::move(dropped)};
 }
 
-struct [[nodiscard]] user_types_to_drop final {
-    seastar::noncopyable_function<future<> ()> drop;
-};
-
 // see the comments for merge_keyspaces()
-static future<user_types_to_drop> merge_types(distributed<service::storage_proxy>& proxy, schema_result before, schema_result after)
+static future<affected_user_types> merge_types(distributed<service::storage_proxy>& proxy, schema_result before, schema_result after)
 {
     auto diff = diff_rows(before, after);
-
-    // Create and update user types before any tables/views are created that potentially
-    // use those types. Similarly, defer dropping until after tables/views that may use
-    // some of these user types are dropped.
-
-    co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) -> future<> {
-        auto created_types = co_await create_types(db, diff.created);
-        for (auto&& user_type : created_types) {
+    affected_user_types affected{
+        .per_shard{smp::count},
+    };
+    co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) mutable -> future<> {
```

Comment:
If affected_user_types is a `sharded\<thing>`, then passing it as an argument to invoke_on_all will convert it to a `thing&`.